pkg/oc/cli/cmd/push/binary.go

```diff
@@ -0,0 +1,363 @@
+	return o.PrintObject([]*resource.Info{info})
+}
+
+func contentsForPathOrURL(s string, in io.Reader, subpaths ...string) (string, io.Reader, int64, error) {
+	switch {
+	case s == "-":
+		return "", in, -1, nil
+	case strings.Index(s, "http://") == 0 || strings.Index(s, "https://") == 0:
+		_, err := url.Parse(s)
+		if err != nil {
+			return "", nil, 0, fmt.Errorf("the URL passed to filename %q is not valid: %v", s, err)
+		}
+		res, err := http.Get(s)
+		if err != nil {
+			return "", nil, 0, err
+		}
+		return "", res.Body, -1, nil
+	default:
+		stat, err := os.Stat(s)
+		if err != nil {
+			return "", nil, 0, err
+		}
+		if !stat.IsDir() {
+			contents, err := os.Open(s)
+			return s, contents, stat.Size(), err
+		}
+		for _, sub := range subpaths {
+			path := filepath.Join(s, sub)
+			stat, err := os.Stat(path)
+			if err != nil {
+				continue
+			}
+			if stat.IsDir() {
+				continue
+			}
+			contents, err := os.Open(path)
+			return path, contents, stat.Size(), err
+		}
+		return s, nil, 0, os.ErrNotExist
+	}
+}
+
+func convertStreamToTarGz(path string, r io.Reader, size int64) io.Reader {
+	br := bufio.NewReaderSize(r, 2048)
+	buf, err := br.Peek(2048)
+	if err != nil && err != io.EOF {
+		glog.V(4).Infof("Got error peeking: %v", err)
+		return errReader{err: err}
+	}
+	switch contentType := http.DetectContentType(buf); contentType {
+	case "application/x-gzip":
+		gr, err := gzip.NewReader(bytes.NewBuffer(buf))
+		if err != nil {
+			glog.V(4).Infof("Not a tar.gz file: %v", err)
+			return errReader{err: fmt.Errorf("%s does not appear to be a valid tar.gz file: %v", path, err)}
+		}
+		tr := tar.NewReader(gr)
+		_, err = tr.Next()
+		if err != nil && err != io.EOF {
+			glog.V(4).Infof("Not a tar.gz file: %v", err)
+			return errReader{err: fmt.Errorf("%s does not appear to be a valid tar.gz file: %v", path, err)}
+		}
+		glog.V(4).Infof("Detected input stream as tar.gz")
+		return br
+	case "application/zip":
+		if size == -1 {
+			glog.V(4).Infof("Not a valid zip file: %v", err)
+			return errReader{err: fmt.Errorf("zip files are only supported from the filesystem", path)}
+		}
+		rAt, ok := r.(io.ReaderAt)
+		if !ok {
+			glog.V(4).Infof("Not a valid zip file: %v", err)
+			return errReader{err: fmt.Errorf("zip files are only supported from the filesystem", path)}
+		}
+		zr, err := zip.NewReader(rAt, size)
+		if err != nil {
+			glog.V(4).Infof("Not a valid zip file: %v", err)
+			return errReader{err: err}
+		}
+		pr, pw := io.Pipe()
+		gw := gzip.NewWriter(pw)
+		tw := tar.NewWriter(gw)
+		go func() {
+			if err := zipToTarGz(zr, tw); err != nil {
+				glog.V(4).Infof("Failed to write to tar.gz: %v", err)
+				pw.CloseWithError(err)
+				return
+			}
+			if err := tw.Close(); err != nil {
+				glog.V(4).Infof("Failed to close tar.gz: %v", err)
+				pw.CloseWithError(err)
+				return
+			}
+			gw.Close()
+			pw.Close()
+		}()
+		glog.V(4).Infof("Detected input stream as zip")
+		return pr
+	default:
+		tr := tar.NewReader(br)
```

Comment:
The call `tr.Next()` will read some data from `br` and that data will not appear in the result stream. It should be `tar.NewReader(bytes.NewBuffer(buf))`.